pkg/controller/snapshot_controller_base.go

```diff
@@ -0,0 +1,468 @@
+
+	ctrl.classLister = volumeSnapshotClassInformer.Lister()
+	ctrl.classListerSynced = volumeSnapshotClassInformer.Informer().HasSynced
+
+	return ctrl
+}
+
+func (ctrl *csiSnapshotController) Run(workers int, stopCh <-chan struct{}) {
+	defer ctrl.snapshotQueue.ShutDown()
+	defer ctrl.contentQueue.ShutDown()
+
+	glog.Infof("Starting CSI snapshotter")
+	defer glog.Infof("Shutting CSI snapshotter")
+
+	if !cache.WaitForCacheSync(stopCh, ctrl.snapshotListerSynced, ctrl.contentListerSynced) {
+		glog.Errorf("Cannot sync caches")
+		return
+	}
+
+	ctrl.initializeCaches(ctrl.snapshotLister, ctrl.contentLister)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.snapshotWorker, 0, stopCh)
+		go wait.Until(ctrl.contentWorker, 0, stopCh)
+	}
+
+	<-stopCh
+}
+
+// enqueueSnapshotWork adds snapshot to given work queue.
+func (ctrl *csiSnapshotController) enqueueSnapshotWork(obj interface{}) {
+	// Beware of "xxx deleted" events
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+	if snapshot, ok := obj.(*crdv1.VolumeSnapshot); ok {
+		objName, err := cache.DeletionHandlingMetaNamespaceKeyFunc(snapshot)
+		if err != nil {
+			glog.Errorf("failed to get key from object: %v, %v", err, snapshot)
+			return
+		}
+		glog.V(5).Infof("enqueued %q for sync", objName)
+		ctrl.snapshotQueue.Add(objName)
+	}
+}
+
+// enqueueContentWork adds snapshot content to given work queue.
+func (ctrl *csiSnapshotController) enqueueContentWork(obj interface{}) {
+	// Beware of "xxx deleted" events
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+	if content, ok := obj.(*crdv1.VolumeSnapshotContent); ok {
+		objName, err := cache.DeletionHandlingMetaNamespaceKeyFunc(content)
+		if err != nil {
+			glog.Errorf("failed to get key from object: %v, %v", err, content)
+			return
+		}
+		glog.V(5).Infof("enqueued %q for sync", objName)
+		ctrl.contentQueue.Add(objName)
+	}
+}
+
+// snapshotWorker processes items from snapshotQueue. It must run only once,
+// syncSnapshot is not assured to be reentrant.
+func (ctrl *csiSnapshotController) snapshotWorker() {
+	workFunc := func() bool {
+		keyObj, quit := ctrl.snapshotQueue.Get()
+		if quit {
+			return true
+		}
+		defer ctrl.snapshotQueue.Done(keyObj)
+		key := keyObj.(string)
+		glog.V(5).Infof("snapshotWorker[%s]", key)
+
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		glog.V(5).Infof("snapshotWorker: snapshot namespace [%s] name [%s]", namespace, name)
+		if err != nil {
+			glog.V(4).Infof("error getting namespace & name of snapshot %q to get snapshot from informer: %v", key, err)
+			return false
+		}
+		snapshot, err := ctrl.snapshotLister.VolumeSnapshots(namespace).Get(name)
+		if err == nil {
+			// The volume snapshot still exists in informer cache, the event must have
+			// been add/update/sync
+			if ctrl.shouldProcessSnapshot(snapshot) {
```

Comment:
here shouldProcessSnapshot might update the snapshot, then this function should return the new version of the snapshot so that the following updateSnapshot can use the new version

Reply:
Fixed.